### PR TITLE
BEEFY: client support for detecting equivocated votes

### PR DESCRIPTION
**Pull request by acatangiu**
Split off the client/worker part of detecting equivocations from https://github.com/paritytech/substrate/pull/13121 into this smaller dedicated PR for easier reviewing.

If we merge this PR and https://github.com/paritytech/substrate/pull/13215 in the same release window, then we can drop the aux-db `PersistentState` migration and rid us of a nice amount of code (because we plan on "restarting" BEEFY consensus during upgrade from "Lean BEEFY" to "Full BEEFY" and old aux-db state will be ignored).

### CODE REVIEW

File: client/beefy/src/aux_schema.rs [hunk 1]
```diff
@@ -72,6 +73,118 @@ where
 	Ok(None)
 }
 
+mod v1 {
```
Comment (acatangiu) [inline client/beefy/src/aux_schema.rs:76]: we can drop this whole migration if we merge the PR in the same release cycle as https://github.com/paritytech/substrate/pull/13215

**Review by svyatonik — APPROVED** [review]

File: primitives/beefy/src/lib.rs [hunk 2]
```diff
@@ -209,6 +209,83 @@ sp_api::decl_runtime_apis! {
 	}
 }
 
+#[cfg(feature = "std")]
```
Comment (andresilva) [inline primitives/beefy/src/lib.rs:212]: Why was this moved to primitives? Maybe it's used in the rest of the PR in runtime code and so that doesn't show here. But ideally this would be compiled for tests, since it's testing infrastructure.

File: client/beefy/src/round.rs [hunk 3]
```diff
@@ -63,27 +57,37 @@ pub fn threshold(authorities: usize) -> usize {
...
 /// Keeps track of all voting rounds (block numbers) within a session.
 /// Only round numbers > `best_done` are of interest, all others are considered stale.
 ///
 /// Does not do any validation on votes or signatures, layers above need to handle that (gossip).
 #[derive(Debug, Decode, Encode, PartialEq)]
-pub(crate) struct Rounds<Payload, B: Block> {
-	rounds: BTreeMap<(Payload, NumberFor<B>), RoundTracker>,
+pub(crate) struct Rounds<B: Block> {
+	rounds: BTreeMap<Commitment<NumberFor<B>>, RoundTracker>,
+	previous_votes: BTreeMap<(Public, NumberFor<B>), VoteMessage<NumberFor<B>, Public, Signature>>,
```
Comment (andresilva) [inline client/beefy/src/round.rs:76]: Not that it matters much for this PR, but I think I would have instead stored this as `Map<NumberFor<B>, Set<VoteMessage<...>>`. The `VoteMessage` already includes the public key, and I was thinking there might be situations in the future where we want to grab all votes for a round. Also this doesn't allow us to store equivocated votes (but maybe you worked around that a different way).

File: client/beefy/src/round.rs [hunk 4]
```diff
@@ -111,60 +115,77 @@ where
...
 		&mut self,
-		round: &(P, NumberFor<B>),
-		vote: (Public, Signature),
-		self_vote: bool,
-	) -> bool {
-		let num = round.1;
+		vote: VoteMessage<NumberFor<B>, AuthorityId, Signature>,
+	) -> VoteImportResult<B> {
+		let num = vote.commitment.block_number;
+		let equivocation_key = (vote.id.clone(), num);
```
Comment (andresilva) [inline client/beefy/src/round.rs:123]: nitpick, but maybe `vote_key` is more appropriate?

File: client/beefy/src/round.rs [hunk 5]
```diff
@@ -111,60 +115,77 @@ where
...
+			.map(|authority_id| votes.get(authority_id).cloned())
+			.collect();
+		SignedCommitment { commitment: round.0, signatures }
 	}
 
 	pub(crate) fn conclude(&mut self, round_num: NumberFor<B>) {
 		// Remove this and older (now stale) rounds.
-		self.rounds.retain(|&(_, number), _| number > round_num);
+		self.rounds.retain(|commitment, _| commitment.block_number > round_num);
+		self.previous_votes.retain(|&(_, number), _| number > round_num);
```
Comment (andresilva) [inline client/beefy/src/round.rs:188]: Not sure if in the future we might want to keep these around for longer. But let's keep it like this for now.

**Review by andresilva — APPROVED** [review]
lgtm. just minor suggestions.

File: primitives/beefy/src/lib.rs (hunk 2 above)
Comment (acatangiu) [reply primitives/beefy/src/lib.rs:212]: Yes, this is later used in `frame/beefy` tests as well as `client/beefy` tests - `primitives/beefy` seemed the obvious place for sharing it.

File: client/beefy/src/round.rs (hunk 3 above)
Comment (acatangiu) [reply client/beefy/src/round.rs:76]: `previous_votes: Map<NumberFor<B>, Set<VoteMessage<...>>>,` would involve iterating over `Set` to look for a possibly equivocated `vote' != vote` (where `vote'` comes from same authority, but with a different commitment).

We could do `Map<NumberFor<B>, Map<Public, VoteMessage<...>>>` to keep same efficient lookup but still allow grabbing all votes for a round.

I say we leave it like this for now, and make the change when/if needed.
